src/test/groovy/pl/allegro/tech/opel/OpelEngineSpec.groovy

```diff
@@ -749,25 +749,150 @@ xyz'"""
+        value.get() == expected
+
+        where:
+        expression                          | expected
+        '2 + myVar'                         | 204
+        '3 + myFunc()'                      | 106
+    }
+
+    @Unroll
+    def "should evaluate expression with variable and function registered in engine with empty context given"() {
+        given:
+        def engine = new OpelEngine()
+        engine.registerCompletedVariable('myVar', 202)
+        engine.registerFunction('myFunc', constFunctionReturning(103))
+
+        when:
+        def value = engine.eval(expression, EvalContext.empty())
+
+        then:
+        value.get() == expected
+
+        where:
+        expression                          | expected
+        '2 + myVar'                         | 204
+        '3 + myFunc()'                      | 106
+    }
+
+    @Unroll
+    def "should prefere variables and functions from eval context"() {
+        given:
+        def engine = new OpelEngine()
+        engine.registerCompletedVariable('myVar', 1)
+        engine.registerFunction('myFunc', constFunctionReturning(2))
+
+        def evalContext = EvalContext.Builder.create()
+                .withFunction('myFunc', constFunctionReturning(1000))
+                .withCompletedVariable('myVar', 2000)
+                .build()
+
+        when:
+        def value = engine.eval(expression, evalContext)
+
+        then:
+        value.get() == expected
+
+        where:
+        expression                          | expected
+        '2 + myVar'                         | 2002
+        '3 + myFunc()'                      | 1003
+    }
+
+    @Unroll
+    def "should evaluate parsed expression with variable and function registered in engine"() {
+        given:
+        def engine = new OpelEngine()
+        engine.registerCompletedVariable('myVar', 202)
+        engine.registerFunction('myFunc', constFunctionReturning(103))
+
+        when:
+        def value = engine.parse(expression).eval()
+
+        then:
+        value.get() == expected
+
+        where:
+        expression                          | expected
+        '2 + myVar'                         | 204
+        '3 + myFunc()'                      | 106
+    }
+
+    @Unroll
+    def "should evaluate parsed expression with variable and function registered in engine with empty context given"() {
+        given:
+        def engine = new OpelEngine()
+        engine.registerCompletedVariable('myVar', 202)
+        engine.registerFunction('myFunc', constFunctionReturning(103))
+
+        when:
+        def value = engine.parse(expression).eval(EvalContext.empty())
+
+        then:
+        value.get() == expected
+
+        where:
+        expression                          | expected
+        '2 + myVar'                         | 204
+        '3 + myFunc()'                      | 106
+    }
+
+    @Unroll
+    def "should prefere variables and functions from eval context when evaluating parsed expression"() {
```

Comment:
I'm not sure, I'm testing that eval context is in use here, not if it was created correctly